src/buildercore/core.py

```diff
@@ -257,20 +257,21 @@ def stack_all_ec2_nodes(stackname, workfn, username=config.DEPLOY_USER, concurre
     ensure(None not in public_ips.values(), "Public ips are not valid: %s", public_ips, exception_class=NoPublicIps)
 
     def single_node_work():
-        try:
-            return workfn(**work_kwargs)
-        except NetworkError as err:
-            if str(err.message).startswith('Timed out trying to connect'):
-                LOG.info("Timeout while executing task on a %s node (%s), retrying once on this node", stackname, err.message)
+        for attempt in range(0, 6):
+            try:
                 return workfn(**work_kwargs)
-            else:
-                raise err
-        except config.FabricException as err:
-            LOG.error(str(err).replace("\n", "    "))
-            # not useful to specify more, this will be printed out
-            # but we are already logging it (and printing it on stderr)
-            # which is better
-            raise SystemExit("")
+            except NetworkError as err:
+                if str(err.message).startswith('Timed out trying to connect'):
+                    LOG.info("Timeout while executing task on a %s node (%s) during attempt %s, retrying on this node", stackname, err.message, attempt)
+                    continue
+                else:
+                    raise err
+            except config.FabricException as err:
+                LOG.error(str(err).replace("\n", "    "))
+                # not useful to specify more, this will be printed out
+                # but we are already logging it (and printing it on stderr)
+                # which is better
+                raise SystemExit("")
```

Comment:
https://docs.python.org/2/library/exceptions.html#exceptions.SystemExit

you can specify an exit code. would a simple integer `1` be better than an empty string here?

Reply:
I'm not sure how we got here, it seems the integer will make it behave differently than the string. I am not sure if this is indeed being unhandled or if Fabric is catching it (and it it can)

Reply:
Experimented. It appears Fabric ignores throwing a `SystemExit` from here, perhaps because we are in a context of a single host while commands are executing in parallel so it has to merge that error with the result of the other hosts.

Reply:
The original commit reports that this is only to silence Fabric: https://github.com/elifesciences/builder/commit/1316ab89d58275cf8e2f3f99945909b43c4bd08b